### DISCUSSION TITLE
Retry when timing-out question

### DISCUSSION THREAD

**Discussion by patjakubik** [sindresorhus/got#1614] [Q&A] [answered]
#### What would you like to discuss?

I'm dealing with quite unresponsive API and would like to retry request up to 3 times if response is taking longer than 1000 ms. 

For that, I'm using options:

```
timeout: 1000,
retry: 3,
```

However, when I inspect requests with my instrumenting tool, I can see that first request took more than 2000 ms.
![image](https://user-images.githubusercontent.com/37963339/108108244-aa517a00-7090-11eb-9b1b-5590e122b112.png)

I would expect that a second request would occur after 1000 ms.

What I'm missing here? Are my options not valid for what I want to achieve?

#### Checklist

- [x] I have read the documentation.

**Comment by szmarczak** [comment, marked as answer]
> Delays between retries counts with function 1000 * Math.pow(2, retry - 1) + Math.random() * 100, where retry is attempt number (starts from 1).

https://github.com/sindresorhus/got#retry

**Reply by patjakubik** [reply]
@szmarczak So delays between first request is ~1000 ms, second ~2000 ms, third ~4000 ms. What's the rationale behind that? How can I achieve what I wrote in the post? Are there any limitations? Sorry but I find that documentation bit enigmatic.

**Reply by szmarczak** [reply]
> How can I achieve what I wrote in the post?

Read the docs. Use `calculateDelay`.

**Reply by szmarczak** [reply]
`({computedValue}) => Math.min(computedValue, 1000)`. It's obvious.

**Reply by szmarczak** [reply]
Or, if you want to retry immediately, use `1` instead.

**Reply by patjakubik** [reply]
Thanks,
do you have an idea why `limit` is not respected here?

```
    got({
      method: "get",
      url: "http://httpstat.us/200?sleep=5000",
      timeout: 1000,
      retry: {
        limit: 3,
        calculateDelay: ({ attemptCount, retryOptions }) => {
          console.log(
            "Retry:",
            { attemptCount, limit: retryOptions.limit },
            new Date().toISOString()
          );
          return 1;
        },
      },
      responseType: "json",
    })
```
    

Returns:
    

```
Retry: { attemptCount: 1, limit: 3 } 2021-02-17T21:17:23.177Z
Retry: { attemptCount: 2, limit: 3 } 2021-02-17T21:17:24.183Z
Retry: { attemptCount: 3, limit: 3 } 2021-02-17T21:17:25.191Z
Retry: { attemptCount: 4, limit: 3 } 2021-02-17T21:17:26.199Z
Retry: { attemptCount: 5, limit: 3 } 2021-02-17T21:17:27.208Z
Retry: { attemptCount: 6, limit: 3 } 2021-02-17T21:17:28.217Z
Retry: { attemptCount: 7, limit: 3 } 2021-02-17T21:17:29.225Z
Retry: { attemptCount: 8, limit: 3 } 2021-02-17T21:17:30.231Z
Retry: { attemptCount: 9, limit: 3 } 2021-02-17T21:17:31.235Z
Retry: { attemptCount: 10, limit: 3 } 2021-02-17T21:17:32.241Z
Retry: { attemptCount: 11, limit: 3 } 2021-02-17T21:17:33.245Z
Retry: { attemptCount: 12, limit: 3 } 2021-02-17T21:17:34.252Z
Retry: { attemptCount: 13, limit: 3 } 2021-02-17T21:17:35.259Z
Retry: { attemptCount: 14, limit: 3 } 2021-02-17T21:17:36.267Z
...and so on
```

I got it sorted out with `return attemptCount < retryOptions.limit ? 1 : 0;` but it would be nice to know if there is better way.

**Reply by szmarczak** [reply]
> do you have an idea why limit is not respected here?

Yes. You didn't handle the `limit` option, therefore it had no effect. It's all in the docs. 

> it would be nice to know if there is better way.

https://github.com/sindresorhus/got/discussions/1614#discussioncomment-377592

**Reply by szmarczak** [reply]
Unless you want something different, then you should make your question more precise.

**Reply by zeing** [reply]
```
let inputStream = got.stream(
    url,
    {
      isStream: true,
      timeout: 500,
      retry: {
        calculateDelay: ({ attemptCount, retryOptions }) => {
          console.log(
            'Retry:',
            { attemptCount, limit: retryOptions.limit },
            new Date().toISOString()
          )
          return 1
        },
      },
    }
  )
```

I have got like 
```
node:events:368
      throw er; // Unhandled 'error' event
      ^

RequestError: Timeout awaiting 'request' for 500ms
```
